source/language/scope.rst

```diff
@@ -0,0 +1,338 @@
+  non-gate and non-subroutine control-flow block, such as ``if`` or ``for``.
+
+* Calibration scopes, which are the shared calibration state contained in
+  ``cal`` and ``defcal`` blocks.  The precise rules for these may vary depending
+  on the particular companion calibration grammar that has been loaded for this
+  program.  See :ref:`the OpenPulse specification <openpulse>` for details on
+  how this is handled in the OpenPulse calibration language.  This document will
+  not cover this topic further.
+
+In general, the lifetime of each identifier begins when it is declared, and ends
+at the completion of the scope it was declared in.  The storage space for each
+variable must be allocated for at least the lifetime of the identifier, but it
+is up to individual implementations to define their allocation strategies within
+nested scopes.
+
+In order to be a valid OpenQASM 3.0 program, symbols must be defined before they
+are used; in particular, there is no forward definition of functions and gates,
+and there can be no mutual recursion.
+
+Most regular identifiers of variables can be shadowed in inner scopes, so that
+the identifier temporarily refers to a different variable, but cannot be
+re-declared without entering an inner scope.  The type of the shadowing variable
+does not need to be the same as the type of the variable it is shadowing.
+
+Not all identifiers declared in outer scopes are visible within inner scopes.
+The visibility depends on the type of the variable, and the type of the inner
+scope.  Approximately, ``const`` variables, gates and subroutines are visible
+within *all* inner scopes, while other variables in outer scopes are visible
+within inner control-flow scopes but not gate and function scopes.
+
+The ``include`` statement should be seen as extending the scope it is contained
+within; all variables that are in scope at the time of the ``include`` statement
+are also in scope while the included file is being parsed, and variables defined
+in that file will be available in the containing scope once the inclusion has
+been parsed.  There is no separate namespacing defined in OpenQASM 3.0.
+
+
+.. _scope-global:
+
+Global scope
+------------
+
+The global scope is active when no other scopes are active.  Several OpenQASM
+3.0 statements are only valid in the global scope, such as (non-exhaustive):
+
+* ``gate``, ``def`` and ``defcal`` declarations,
+* ``qubit`` declarations,
+* ``array`` declarations.
+
+A simple example of scoping between a main program file and an included file:
+
+.. code-block:: c
+   :caption: Main program
+
+   OPENQASM 3.0;
+
+   gate h q {
+      U(pi/2, 0, pi) q;
+      // `U` is the single-qubit gate that is implicitly defined in the global
+      // scope of all OpenQASM 3.0 programs, and so is available here.
+      // Similarly `pi` is one of the implicitly defined constants.
+   }
+
+   int i = 100;
+
+   include "my_definitions.qasm";
```

Comment:
**thought**: The idea that "my_definitions.qasm" is not a valid QASM was a bit jarring. Is this unavoidable (or already present) in the language as it's now written?

Reply:
[Right now, the spec says](https://openqasm.com/language/comments.html#included-files):

> The statement `include "filename";` continues parsing `filename` as if the contents of the file were inserted at the location of the `include` statement.

That implies that `filename` could be an incomplete OQ 3 program (like the example here).  I think it's worth making a positive decision whether that's exactly what we want, or if we want something different.

I think the two ends of the spectrum that we can define `include` on are:
- textual "copy-paste the contents of this file here" includes, like C's `#include`
- the file must be a completely semantically valid OQ 3 program, starting off with nothing in scope, and anything defined at the end becomes in scope in the containing context.

Since we don't have any visibility control mechanisms (namespacing / some sort of `public`/`private` keywords), any `include` system has issues bringing in the symbols it needs without causing clashes with other libraries.  You can't have `library_a.qasm` and `library_b.qasm` co-exist in the same file if they both `include "stdgates.inc";`, because they'll redefine all the symbols, and that's forbidden in OQ 3.

A work-around would be to say that the _user_ is responsible for `include`ing all the dependencies themselves, but that's not good extensible design for a language.  Some languages solve this with namespacing, some with explicitly marking the symbols made public from a file, and some with preprocessor conditional `include` guards.  Right now, we don't have the mechanisms for any of those solutions, so we're kind of requiring the "user supplies the dependencies" method.

Reply:
Is there any use case for having `include` anywhere other than global scope? For the common case of including some gate/extern defs it is only at global scope. I think 99% of use cases could be captured by saying you can only have an `include` at global scope and the included file must end at global scope (ie it shouldn't end half-way through a block). This wouldn't preclude nested includes (`main` includes `lib` that in turn includes `module`).

Namespacing is probably a good idea for a future enhancement something like `include "special_gates" prefix abc;` making all the global symbols introduced within the included file available with a prefix `abc::special_cnot` etc but i think that can be added backward compatible with the above

Reply:
I agree it seems pretty useless anywhere but global scope (and the reference parser already rejects `include` in non-global scopes).  The problems of leaking gate definitions is independent of that, though.  As a more complete example, consider two files `library_a.qasm`:
```qasm
include "stdgates.inc";

gate gate_a q {
    x q;
}

// at this point, `gate_a` is in scope and will be exported,
// but so is everything defined in `stdgates.inc`.
```
and `library_b.qasm`:
```qasm
include "stdgates.inc";

gate gate_b q {
    z q;
}
```

So far this is ok, but now if I have my program:
```qasm
OPENQASM 3.0;

include "library_a.qasm";
// ok.
include "library_b.qasm";
// Error: cannot redefine {x, y, z, h, cx, ...} (everything in stdgates.inc).
```

Currently, the way around this is for neither library to actually `include` its dependencies, and to rely on the user to do that exactly once.  That requires that `include` statements start with all the outer identifiers in scope, which is how the current text is written.

There's a few other potential ways around that, including:
- some form of header-guard syntax (`#if !defined(_STDGATES_INC)`)
- some form of namespacing
- some form of marking what symbols should be exported from a file/scope (such as a `public` keyword)

Reply:
From TSC 2022-06-10: we acknowledge that these effects aren't ideal, but there _is_ a workaround, and fixing it is out of scope for this PR.  We don't immediately agree on the details of an alternative, so we don't necessarily need to fix it for 3.0.

Reply:
`include` is explicitly stated as global in 01ed7cc per the TSC decision.  I'll leave the rest of this comment chain for context in the future.

Reply:
I've moved the context of this comment chain into its own issue in #365, so I'll mark this as resolved for this PR, since we decided not to make any changes here.